File: content/snippets/go/design_patterns_abstract_factory.go
package main

import "fmt"

type Button interface{ Render() string }
type Dialog interface{ Render() string }

type WidgetFactory interface {
	CreateButton() Button
	CreateDialog() Dialog
}

type LightButton struct{}

func (LightButton) Render() string { return "light button" }

type LightDialog struct{}

func (LightDialog) Render() string { return "light dialog" }

type DarkButton struct{}

func (DarkButton) Render() string { return "dark button" }

type DarkDialog struct{}

func (DarkDialog) Render() string { return "dark dialog" }

type LightFactory struct{}

func (LightFactory) CreateButton() Button { return LightButton{} }
func (LightFactory) CreateDialog() Dialog { return LightDialog{} }

type DarkFactory struct{}

func (DarkFactory) CreateButton() Button { return DarkButton{} }
func (DarkFactory) CreateDialog() Dialog { return DarkDialog{} }

func showScreen(factory WidgetFactory) {
	button := factory.CreateButton()
	dialog := factory.CreateDialog()
	fmt.Printf("%s + %s\n", button.Render(), dialog.Render())
}

func main() {
	showScreen(LightFactory{})
	showScreen(DarkFactory{})
}

File: content/snippets/typescript/design_patterns_abstract_factory.ts
interface Button {
  render(): string;
}

interface Dialog {
  render(): string;
}

interface WidgetFactory {
  createButton(): Button;
  createDialog(): Dialog;
}

class LightButton implements Button {
  render(): string {
    return 'light button';
  }
}

class LightDialog implements Dialog {
  render(): string {
    return 'light dialog';
  }
}

class DarkButton implements Button {
  render(): string {
    return 'dark button';
  }
}

class DarkDialog implements Dialog {
  render(): string {
    return 'dark dialog';
  }
}

class LightFactory implements WidgetFactory {
  createButton(): Button {
    return new LightButton();
  }

  createDialog(): Dialog {
    return new LightDialog();
  }
}

class DarkFactory implements WidgetFactory {
  createButton(): Button {
    return new DarkButton();
  }

  createDialog(): Dialog {
    return new DarkDialog();
  }
}

function showScreen(factory: WidgetFactory): void {
  console.log(`${factory.createButton().render()} + ${factory.createDialog().render()}`);
}

showScreen(new LightFactory());
showScreen(new DarkFactory());